Timer envelope removed for AutoReceivedMessage (#24080)

mucahitkantepe opened the pull request:
#24080 

Timers wraps any message(including `AutoReceivedMessage`'s).

`akka.actor.ActorCell.invoke` method acts like it is not an `AutoReceivedMessage` and calls `aroundReceive()` method that applies `receive` method.

As a result old implementation will not handle `AutoReceivedMessage` automatically, it applies custom receive method.

This PR bypasses `akka/actor/Timers.scala:38` case, **solution might be problematic.**

patriknw commented on akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala:
```diff
@@ -45,7 +45,10 @@ import akka.util.OptionVal
 
-    val timerMsg = TimerMsg(key, nextGen, this)
+    val timerMsg = msg match {
+      case _: AutoReceivedMessage ⇒ msg
```
This will not work. Then all checks that are done when receiving a TimerMsg is not done. I think nothing should be done here and instead in `Timers.aroundReceive` add something like:

```
          case OptionVal.Some(m: AutoReceivedMessage) ⇒
            self.asInstanceOf[LocalActorRef].underlying.autoReceiveMessage(Envelope(m, sender()))
```

It would be good to add a test also.

mucahitkantepe replied:
Your approach works for Actor's that composed with `Timers` trait.

Also same bug appears in `akka.actor.FSM` class.  `akka.actor.FSM.setTimer` method creates a `akka.actor.FSM.Timer` and calls `schedule` method and it wraps `AutoReceivedMessage` with `akka.actor.FSM.Timer`. So FSM Actor actually receives `PoisonPill` or any other AutoReceivedMessage.

Possible solution is again using  in `akka.actor.FSM.Timer.schedule` method, something like:

``` 
val timerMsg = msg match {
        case m: AutoReceivedMessage ⇒ m
        case _                      ⇒ this
      }
``` 
Is it convenient for you?

patriknw replied:
@mucahitkantepe yes, it's fine to do similar fix in FSM

patriknw replied:
and `PersistentFSM` :(

patriknw reviewed (changes requested):
Great finding. Must be implemented in another way...

mucahitkantepe committed:
AutoReceivedMessage with Timers bug fixed (#24080)

AutoReceivedMessage with FSM schedule bug fixed (#24080)

Tests added for both FSM and Timers trait for (#24080)

AutoReceivedMessage with PersistentFSM bug fixed and test added (#24080)

patriknw reviewed (commented):
@mucahitkantepe solution is looking good, please rebase to fix the merge conflict (you can squash to one commit if you like)

patriknw commented:
fixed merge conflict

patriknw reviewed (approved):
LGTM

patriknw commented:
don't squash and merge, I have commit here
I'll squash manually.

patriknw commented:
I have squashed this properly

ktoso reviewed (approved):
LGTM
